src/package/packageVersionRetrieve.ts

```diff
@@ -0,0 +1,244 @@
+    type: 'directory',
+    outputDirectory: destinationPath,
+    genUniqueDir: false,
+  });
+
+  await attemptToUpdateProjectJson(
+    project,
+    connection,
+    versionInfo.MetadataPackageId,
+    subscriberPackageVersionId,
+    dependencies,
+    destinationFolder
+  );
+
+  return result;
+}
+
+/**
+ * Attempt to update the sfdx-project.json file to add information about the retrieved sources. If this fails for some reason we
+ * print out an error message and return so the user will still see a list of retrieved metadata.
+ *
+ */
+async function attemptToUpdateProjectJson(
+  project: SfProject,
+  connection: Connection,
+  packageId: string,
+  subscriberPackageVersionId: string,
+  dependencyIds: string[],
+  destinationFolder: string
+): Promise<void> {
+  if (process.env.SF_PROJECT_AUTOUPDATE_DISABLE_FOR_PACKAGE_VERSION_RETRIEVE) {
+    getLogger().info(
+      'Skipping sfdx-project.json updates because SF_PROJECT_AUTOUPDATE_DISABLE_FOR_PACKAGE_VERSION_RETRIEVE is set'
+    );
+    return;
+  }
+  try {
+    const packageInfo: PackagingSObjects.MetadataPackage = (await connection.tooling
+      .sobject('MetadataPackage')
+      .retrieve(packageId)) as PackagingSObjects.MetadataPackage;
+
+    if (packageInfo.PackageCategory !== 'Package2') {
+      getLogger().info(
+        `Skipping sfdx-project.json updates because ${packageId} is not a 2GP package. It has a PackageCategory of '${packageInfo.PackageCategory}'`
+      );
+      return;
+    }
+
+    const queryOptions = {
+      whereClause: `WHERE SubscriberPackageVersionId = '${subscriberPackageVersionId}'`,
+    };
+
+    const versions = await PackageVersion.queryPackage2Version(connection, queryOptions);
+
+    if (versions.length && versions[0]) {
+      const version = versions[0];
+      const pkg = new Package({
+        packageAliasOrId: version.Package2Id,
+        project,
+        connection,
+      });
+
+      const pkgData = await pkg.getPackageData();
+      if (pkgData) {
+        const dirEntry = createPackageDirEntry(project, {
+          name: pkgData.Name,
+          description: pkgData.Description,
+          path: destinationFolder,
+          noNamespace: pkgData.NamespacePrefix != null,
+          orgDependent: pkgData.IsOrgDependent,
+          packageType: pkgData.ContainerOptions,
+          errorNotificationUsername: pkgData.PackageErrorUsername,
+        });
+
+        // Dependencies listed in a package dir need a 'package' (name), which we don't have in the retrieved zip.
+        // We only have a list of 04t... ids so we will make a couple of queries to look up the name of each dependency.
+        const dependencies: PackageDirDependency[] = [];
+        for (const dep of dependencyIds) {
+          // eslint-disable-next-line no-await-in-loop
+          const name = await getPackageNameForSubscriberPackageVersionId(connection, dep);
+          dependencies.push({
+            package: name,
```

Comment:
This probably works, but I usually see either: 
`"dependencies": [
                {
                    "package": "04txx0000004JDC"
                }
            ]` or 
`"dependencies": [ 
   {
      "package": "MyPackageName@0.1.0.1"
   }
]`

Reply:
Or maybe the PackageDirDependency converts it? Just a comment...

Reply:
Using "package": "04t..." works and will clean up the implementation a bit. I'll update the review.